swarm-test/src/lib.rs

```diff
@@ -0,0 +1,353 @@
+///
+/// Using [`drive`] instead of [`futures::future::join`] ensures that a [`Swarm`] continues to be polled, even after it emitted its events.
+pub async fn drive<
+    TBehaviour1,
+    const NUM_EVENTS_SWARM_1: usize,
+    Out1,
+    TBehaviour2,
+    const NUM_EVENTS_SWARM_2: usize,
+    Out2,
+>(
+    swarm1: &mut Swarm<TBehaviour2>,
+    swarm2: &mut Swarm<TBehaviour1>,
+) -> ([Out1; NUM_EVENTS_SWARM_1], [Out2; NUM_EVENTS_SWARM_2])
+where
+    TBehaviour2: NetworkBehaviour + Send,
+    TBehaviour2::OutEvent: Debug,
+    TBehaviour1: NetworkBehaviour + Send,
+    TBehaviour1::OutEvent: Debug,
+    SwarmEvent<TBehaviour2::OutEvent, THandlerErr<TBehaviour2>>: TryIntoOutput<Out1>,
+    SwarmEvent<TBehaviour1::OutEvent, THandlerErr<TBehaviour1>>: TryIntoOutput<Out2>,
+    Out1: Debug,
+    Out2: Debug,
+{
+    let mut res1 = Vec::<Out1>::with_capacity(NUM_EVENTS_SWARM_1);
+    let mut res2 = Vec::<Out2>::with_capacity(NUM_EVENTS_SWARM_2);
+
+    while res1.len() < NUM_EVENTS_SWARM_1 || res2.len() < NUM_EVENTS_SWARM_2 {
+        match futures::future::select(swarm1.next_swarm_event(), swarm2.next_swarm_event()).await {
+            Either::Left((o1, _)) => {
+                if let Ok(o1) = o1.try_into_output() {
+                    res1.push(o1);
+                }
+            }
+            Either::Right((o2, _)) => {
+                if let Ok(o2) = o2.try_into_output() {
+                    res2.push(o2);
+                }
+            }
+        }
+    }
+
+    (
+        res1.try_into().unwrap_or_else(|res1: Vec<_>| {
+            panic!(
+                "expected {NUM_EVENTS_SWARM_1} items from first swarm but got {}",
+                res1.len()
+            )
+        }),
+        res2.try_into().unwrap_or_else(|res2: Vec<_>| {
+            panic!(
+                "expected {NUM_EVENTS_SWARM_2} items from second swarm but got {}",
+                res2.len()
+            )
+        }),
+    )
+}
+
+pub trait TryIntoOutput<O>: Sized {
+    fn try_into_output(self) -> Result<O, Self>;
+}
+
+impl<O, THandlerErr> TryIntoOutput<O> for SwarmEvent<O, THandlerErr> {
+    fn try_into_output(self) -> Result<O, Self> {
+        self.try_into_behaviour_event()
+    }
+}
+impl<TBehaviourOutEvent, THandlerErr> TryIntoOutput<SwarmEvent<TBehaviourOutEvent, THandlerErr>>
+    for SwarmEvent<TBehaviourOutEvent, THandlerErr>
+{
+    fn try_into_output(self) -> Result<SwarmEvent<TBehaviourOutEvent, THandlerErr>, Self> {
+        Ok(self)
+    }
+}
+
+#[async_trait]
+impl<B> SwarmExt for Swarm<B>
+where
+    B: NetworkBehaviour + Send,
+    <B as NetworkBehaviour>::OutEvent: Debug,
+{
+    type NB = B;
+
+    fn new_ephemeral(behaviour_fn: impl FnOnce(Keypair) -> Self::NB) -> Self
+    where
+        Self: Sized,
+    {
+        let identity = Keypair::generate_ed25519();
+        let peer_id = PeerId::from(identity.public());
+
+        let transport = MemoryTransport::default()
+            .or_transport(libp2p_tcp::async_io::Transport::default())
```

Comment:
Why do we need the TCP transport? Is this needed for `libp2p-autonat`?

Reply:
Yes, and `dcutr` needs it too.